# PROJECT 1 - DATA CLEANING AND PREPARATION
## DecodeLabs Data Analytics Internship

In [1]:
import pandas as pd

# Load dataset
file_path = r"C:\Users\onyer\Downloads\Dataset for Data Analytics.xlsx"
df = pd.read_excel(file_path)

In [2]:
# Dataset Overview
print(df.head(10))
print("\nThe size of the dataset:")
print("Columns:", df.shape[1])
print("Rows:", df.shape[0])

     OrderID       Date CustomerID  Product  Quantity  UnitPrice  \
0  ORD200000 2023-01-04     C72649  Monitor         5     570.62   
1  ORD200001 2024-08-23     C75739    Phone         2     151.35   
2  ORD200002 2024-02-27     C81728   Tablet         5     550.68   
3  ORD200003 2023-10-15     C33540    Chair         1     273.19   
4  ORD200004 2025-05-08     C81840  Printer         4     626.01   
5  ORD200005 2023-10-23     C37249    Phone         2     245.86   
6  ORD200006 2025-06-17     C83492   Laptop         1     664.42   
7  ORD200007 2023-05-12     C41460  Monitor         5     149.55   
8  ORD200008 2025-04-02     C26817    Phone         2     134.28   
9  ORD200009 2023-11-21     C31946     Desk         4     509.38   

  ShippingAddress PaymentMethod OrderStatus TrackingNumber  ItemsInCart  \
0     928 Main St    Debit Card     Shipped    TRK37947903            7   
1     823 Main St        Online     Shipped    TRK91186779            3   
2     512 Main St   Credit

In [3]:
# Missing Values
print("Missing values before cleaning\n", df.isnull().sum())

Missing values before cleaning
 OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64


In [4]:
# Replace missing CuouponCode with "No Coupon"
df["CouponCode"] = df["CouponCode"].fillna("NO COUPON")
print("Missing values after cleaning\n", df.isnull().sum())

Missing values after cleaning
 OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64


In [5]:
# Duplicated values 
print("Duplicated Rows found:", df.duplicated().sum())

Duplicated Rows found: 0


In [6]:
# Check for duplicated CustomerID
print("Duplicated OrderID found:", df["OrderID"].duplicated().sum())

Duplicated OrderID found: 0


In [7]:
# Check for data types
print("Data types\n", df.dtypes)

Data types
 OrderID                       str
Date               datetime64[us]
CustomerID                    str
Product                       str
Quantity                    int64
UnitPrice                 float64
ShippingAddress               str
PaymentMethod                 str
OrderStatus                   str
TrackingNumber                str
ItemsInCart                 int64
CouponCode                    str
ReferralSource                str
TotalPrice                float64
dtype: object


In [8]:
# Converting Date to correct datetime format
df["Date"] = df["Date"].astype("datetime64[ns]")
print("Data types after conversion\n", df.dtypes)

invalid_dates = df["Date"].isnull().sum()
print("\ninvalid_dates:", invalid_dates)

Data types after conversion
 OrderID                       str
Date               datetime64[ns]
CustomerID                    str
Product                       str
Quantity                    int64
UnitPrice                 float64
ShippingAddress               str
PaymentMethod                 str
OrderStatus                   str
TrackingNumber                str
ItemsInCart                 int64
CouponCode                    str
ReferralSource                str
TotalPrice                float64
dtype: object

invalid_dates: 0


In [9]:
# Numerical columns validation
df.select_dtypes(include=["number"]).dtypes

Quantity         int64
UnitPrice      float64
ItemsInCart      int64
TotalPrice     float64
dtype: object

In [10]:
df[["Quantity", "UnitPrice", "ItemsInCart", "TotalPrice"]].isna().sum()

Quantity       0
UnitPrice      0
ItemsInCart    0
TotalPrice     0
dtype: int64

In [11]:
# Checked for trailing whitespace
df = df.apply(lambda col: col.str.strip() if col.dtype == "object" else col)

df["Product"].head()

0    Monitor
1      Phone
2     Tablet
3      Chair
4    Printer
Name: Product, dtype: str

In [12]:
# check for outlier
print(df.describe())

                      Date     Quantity    UnitPrice  ItemsInCart   TotalPrice
count                 1200  1200.000000  1200.000000  1200.000000  1200.000000
mean   2024-03-22 16:58:48     2.945833   356.412750     5.485000  1053.968300
min    2023-01-01 00:00:00     1.000000    11.390000     1.000000    11.390000
25%    2023-08-03 18:00:00     2.000000   186.062500     4.000000   410.520000
50%    2024-03-23 00:00:00     3.000000   364.210000     5.000000   823.615000
75%    2024-11-08 12:00:00     4.000000   521.570000     7.000000  1578.475000
max    2025-06-30 00:00:00     5.000000   699.930000    10.000000  3456.400000
std                    NaN     1.407557   197.177146     2.281983   819.856558


In [13]:
# Copy of the cleaned dataset
df_clean = df.copy()
print(df_clean.head())

     OrderID       Date CustomerID  Product  Quantity  UnitPrice  \
0  ORD200000 2023-01-04     C72649  Monitor         5     570.62   
1  ORD200001 2024-08-23     C75739    Phone         2     151.35   
2  ORD200002 2024-02-27     C81728   Tablet         5     550.68   
3  ORD200003 2023-10-15     C33540    Chair         1     273.19   
4  ORD200004 2025-05-08     C81840  Printer         4     626.01   

  ShippingAddress PaymentMethod OrderStatus TrackingNumber  ItemsInCart  \
0     928 Main St    Debit Card     Shipped    TRK37947903            7   
1     823 Main St        Online     Shipped    TRK91186779            3   
2     512 Main St   Credit Card   Cancelled    TRK42903982            8   
3     275 Main St    Debit Card    Returned    TRK62788070            5   
4     668 Main St        Online   Delivered    TRK29241424            8   

  CouponCode ReferralSource  TotalPrice  
0     SAVE10      Instagram     2853.10  
1     SAVE10       Referral      302.70  
2   FREESHIP  

In [14]:
df_clean.to_excel("Week-01-Cleaned-Dataset.xlsx", index=False)
print("Save successful as Week-01-Cleaned-Dataset.xlsx")

Save successful as Week-01-Cleaned-Dataset.xlsx


# 📊 Data Cleaning Report

## 📌 Dataset Summary
- Total Records: 1200
- Total Columns: 14

🧹 Cleaning Tasks Performed
| Issue Identified               | Action Taken                                                     | Rows Affected    |
| ------------------------------ | ---------------------------------------------------------------- | ---------------- |
| Missing CouponCode             | Replaced missing values with `"NO COUPON"`                       | 309              |
| Missing values (general check) | Identified using `isna().sum()` and handled where necessary      | Various          |
| Duplicate Records              | No duplicate rows found                                          | 0                |
| Duplicate Order IDs            | No duplicate OrderID found                                       | 0                |
| Data Types                     | Standardized all columns to correct formats                      | 1200             |
| Date Formatting Issues         | Converted Date column to `datetime64[ns]`                        | 1200             |
| Invalid Date Values            | No invalid dates detected                                        | 0                |
| Numeric Validation             | Ensured Quantity, UnitPrice, ItemsInCart, TotalPrice are numeric | 1200             |
| Text Inconsistencies           | Removed extra spaces and standardized text fields                | Multiple columns |

## 🧾 Data Cleaning Steps Performed
- Loaded dataset using pandas from Excel file
- Performed initial inspection using head(), info(), and describe()
- Checked for missing values across all columns
- Replaced missing values in CouponCode with "NO COUPON"
- Verified dataset for duplicates (none found)
- Validated uniqueness of OrderID
- Converted Date column to proper datetime format (datetime64[ns])
- Ensured numeric columns were correctly typed and validated
- Checked for invalid numeric entries using .isna().sum()
- Standardized text fields by removing unnecessary spaces
- Reviewed data distribution using descriptive statistics (describe())
- Performed outlier analysis using min, max, mean, and quartiles
- Confirmed no extreme or invalid outliers present
- Exported cleaned dataset for further analysis

## 🚀 Outcome

The dataset was successfully cleaned and prepared for analysis.
All identified data quality issues were resolved, ensuring:

Accurate and consistent data structure
Reliable numerical calculations
Proper time-series formatting
Clean categorical values
Analysis-ready dataset for visualization and reporting

The dataset is now fully suitable for:

- Exploratory Data Analysis (EDA)  
- Power BI / Tableau dashboards  
- Business insights generation  


## ⭐ Final Note

This cleaning process ensures high data integrity and improves the reliability of all future analysis and decision-making